libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Runtime/Adapters/MultiplexingAdapter.cs

```diff
@@ -0,0 +1,170 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Bot.Builder.Dialogs.Adaptive.Runtime.Settings;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Bot.Builder.Integration.AspNet.Core
+{
+    /// <summary>
+    /// A multiplexing bot adapter that can select the appropriate IbotFrameworkHttpAdapter based on the incoming
+    /// request's route or the activity's ChannelId.
+    /// </summary>
+    /// <remarks>
+    /// To use this adapter properly, the api/{route} must match the Activity.ChannelId to map the appropriate adapter.
+    /// Otherwise, you may override the BuildAdapterMap() method.
+    /// </remarks>
+    public class MultiplexingAdapter : BotAdapter, IBotFrameworkHttpAdapter
+    {
+        private readonly Dictionary<string, IBotFrameworkHttpAdapter> _adapters = new ();
+        private readonly IConfiguration _configuration;
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MultiplexingAdapter"/> class.
+        /// </summary>
+        /// <param name="configuration">An <see cref="IConfiguration"/> instance.</param>
+        /// <param name="adapters">The array of adapters used in multiplexing.</param>
+        /// /// <param name="logger">The ILogger implementation this adapter should use.</param>
+        public MultiplexingAdapter(IConfiguration configuration, IEnumerable<IBotFrameworkHttpAdapter> adapters, ILogger logger = null)
+        {
+            _configuration = configuration;
+            _logger = logger ?? NullLogger.Instance;
+
+            Initialize(adapters);
+        }
+
+        /// <inheritdoc/>
+        public Task ProcessAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken = default)
+        {
+            var routeAsChannelId = httpRequest.Path.ToString().Split('/').LastOrDefault();
+            return (GetAppropriateAdapter(routeAsChannelId) as IBotFrameworkHttpAdapter).ProcessAsync(httpRequest, httpResponse, bot, cancellationToken);
+        }
```

Comment:
This is the only method where we don't actually know the `Activity.ChannelId` yet. However, if we ensure that the route (`/api/{route}`) used exactly matches the `ChannelId`, we can use the route as the ChannelId. The need to match the route to the `ChannelId` is somewhat configurable, since `BuildAdapterMap()` is a `virtual` method.